File: app/src/solver/WordNode.js
var quat = quat || {};
quat.solver = quat.solver || {};

/* A conglomeration of four labels we use to display a word. */
quat.solver.WordNode = cc.Node.extend({

    applyTheme: function(theme) {
        this.setColor(theme.colors.text);
    },

    /**
     * Creates a WordNode with the given font size.
     * @param  {number} fontSize The size of the font to be initialized.
     * @param  {number} fontGap The distance between each letter.
     */
    ctor: function (fontSize, fontGap) {
        this._super();

        // Store these for later
        this.fontSize = fontSize;
        this.fontGap = fontGap;

        var offset = -1 * ((fontGap * 1.5));

        // Create a pool of labels we can use to draw this word
        var letterPool = [];
        for (var j = 0; j < 4; j++) {
            var letterLabel = new cc.LabelTTF("A", "Ubuntu", fontSize);

            // Set them to be hidden
            letterLabel.x = (j * this.fontGap) + offset;
            letterLabel.y = 0;

            // letterLabel.setColor(cc.color(176,196,222,255));
			letterLabel.setColor(cc.color(255,255,255,255));
            // Add it as a child to this layer
            this.addChild(letterLabel);

            letterPool.push(letterLabel);
        }
        this.letterPool = letterPool;

        this.recalculateBounds();
        this.oldX = this.x;
        this.oldY = this.y;

        return true;
    },

    recalculateBounds: function() {
        // Don't recalculate if x hasn't moved
        if ((this.oldX == this.x) && (this.oldY == this.y)) {
            return;
        }

        var x = this.x,
            y = this.y,
            fontHeight = this.fontSize,
            fontHeightHalf = fontHeight / 2,
            fontWidth = fontHeight * 0.8,
            fontWidthHalf = fontWidth / 2,
            bounds = [];

        // Calculate bound rectangles for all the letters
        for (var j = 0; j < 4; j++) {
            var letter = this.letterPool[j];

            // Set them to be hidden
            bounds.push(cc.rect(x - letter.x - fontWidthHalf, 
                                y - fontHeightHalf,
                                fontWidth,
                                fontHeight));
        }

        // Not sure why we need this
        bounds.reverse();

        // Calculate the bound rectangle for the whole word
        var first = bounds[0],
            last = bounds[3],
            whole = cc.rect(first.x, 
                            first.y, 
                            ((last.x + last.width) - first.x),
                            first.height);

        // Store the bounds
        this.bounds = bounds;
        this.wholeBounds = whole;
        this.oldX = x;
        this.oldY = y;
    },

    pointInWord: function(point) {
        var letterPool = this.letterPool;

        this.recalculateBounds();
        
        var bounds = this.bounds;
        for (var j = 0; j < 4; j++) {
            var bound = bounds[j];
            if (cc.rectContainsPoint(bound, point)) {
                return j;
            }
        }

        return false;
    },

    pointInWhole: function(point) {
        return cc.rectContainsPoint(this.wholeBounds, point);
    },

    /**
     * Replaces the text of this word with a given word.
     * @param  {[type]} word [description]
     */
    changeWord: function(word) {
        for (var j = 0; j < 4; j++) {
            var letterLabel = this.letterPool[j];
            letterLabel.string = word[j].toUpperCase();
        }
    },

    getWord: function() {
        var str = "";
        for (var j = 0; j < 4; j++) {
            var letterLabel = this.letterPool[j];
            str += letterLabel.string;
        }
        return str;
    },

    changeLetter: function(column, letter) {
        this.letterPool[column].string = letter;
    },

    getLetter: function(column) {
        return this.letterPool[column].string;
    },

    setColor: function(color) {
        for (var j = 0; j < 4; j++) {
            var letterLabel = this.letterPool[j];
            letterLabel.setColor(color);
            letterLabel.setVisible(false);
            letterLabel.setVisible(true);
        }
    },

    setOpacity: function(opacity) {
        for (var j = 0; j < 4; j++) {
            var letterLabel = this.letterPool[j];
            letterLabel.setOpacity(opacity);
        }
    }
});

File: cpp/Classes/solver/WordNode.cpp
#include "WordNode.h"

#include <vector>
#include <string>

namespace QUAT {

void WordNode::changeLetter(int column, std::string * letter) {
	auto letterLabel = (*this->letterPool)[column];

	// Set the letter label to this letter
	letterLabel->setString(*letter);
}

void WordNode::changeWord(std::string * word) {
	for (int i = 0; i < 4; i++) {
		// Grab the letter in the word string
		std::string sub = word->substr(i, 1);

		auto letterLabel = (*this->letterPool)[i];

		// Set the letter label to this letter
		letterLabel->setString(sub);
	}
}

std::string * WordNode::getLetter(int column) {
	return NULL;
}

std::string * WordNode::getWord() {
	return NULL;
}

bool WordNode::init() {
	// Init the super class
    if ( !Layer::init() )
    {
        return false;
    }
	
	float offset = -1 * (this->fontGap * 1.5);

	// Initialize the whole bounds
	whole = new cocos2d::Rect(0,0,0,0);

	// Initialize the pool of letters
	letterPool = new std::vector<cocos2d::Label*>();

	// Initialize the bounds of the letters
	bounds = new std::vector<cocos2d::Rect*>();

	for (int i = 0; i < 4; i++) {
		auto letterLabel = cocos2d::Label::createWithTTF("A", "fonts/Arimo-Regular.ttf", this->fontSize);

		// Set them to their proper location
		letterLabel->setPositionX((((float) i) * this->fontGap) + offset);
		letterLabel->setPositionY(0);

		// Make them white
		letterLabel->setColor(cocos2d::Color3B(255, 255, 255));

		this->addChild(letterLabel);

		letterPool->push_back(letterLabel);
		bounds->push_back(new cocos2d::Rect(0,0,0,0));
	}


	this->recalculateBounds();

	return true;
}

bool WordNode::pointInWhole(cocos2d::Point * point) {
	return this->whole->containsPoint(*point);
}

int WordNode::pointInWord(cocos2d::Point * point) {
	for (int i = 0; i < 4; i++) {
		auto rect = (*this->bounds)[i];

		if (rect->containsPoint(*point)) {
			return i;
		}
	}

	return -1;
}

void WordNode::recalculateBounds() {
	// Initialize all the pretty calculations
	float x = this->getPositionX(),
		  y = this->getPositionY(),
		  fontHeight = this->fontSize,
		  fontHeightHalf = fontHeight / 2,
		  fontWidth = fontHeight * 0.8,
		  fontWidthHalf = fontWidth / 2;

	// Calculate bound rectangles for all of the letters
	for (int i = 0; i < 4; i++) {
		auto letter = (*this->letterPool)[i];
		auto rect = (*this->bounds)[i];

		rect->setRect(x - letter->getPositionX() - fontWidthHalf,
					  y - fontHeightHalf,
					  fontWidth,
					  fontHeight);
	}

	auto first = (*this->bounds)[0],
		 last  = (*this->bounds)[3];

	whole->setRect(first->origin.x,
				   first->origin.y,
				   ((last->origin.x + last->size.width) - first->origin.x),
				   first->size.height);
}

WordNode::WordNode(float fontSize, float fontGap) {
	this->fontSize = fontSize;
	this->fontGap = fontGap;
}

WordNode * WordNode::create(float fontSize, float fontGap)
{
    WordNode *pRet = new(std::nothrow) WordNode(fontSize, fontGap);
    if (pRet && pRet->init())
    {
        pRet->autorelease();
        return pRet;
    }
    else
    {
        delete pRet;
        pRet = NULL;
        return NULL;
    }
}

}